listenbrainz/listenstore/timescale_listenstore.py

```diff
@@ -179,177 +174,124 @@ def get_timestamps_for_user(self, user_name):
         """ Return the max_ts and min_ts for a given user and cache the result in brainzutils cache
         """
 
-        tss = cache.get(REDIS_USER_TIMESTAMPS % user_name)
+        tss = cache.get(self.ns + REDIS_USER_TIMESTAMPS % user_name)
         if tss:
             (min_ts, max_ts) = tss.split(",")
             min_ts = int(min_ts)
             max_ts = int(max_ts)
         else:
-            query = 'SELECT first(artist_msid) FROM ' + get_escaped_measurement_name(user_name)
-            min_ts = self._select_single_timestamp(query, get_measurement_name(user_name))
+            min_ts = self._select_single_timestamp(True, user_name)
+            max_ts = self._select_single_timestamp(False, user_name)
 
-            query = 'SELECT last(artist_msid) FROM ' + get_escaped_measurement_name(user_name)
-            max_ts = self._select_single_timestamp(query, get_measurement_name(user_name))
-
-            cache.set(REDIS_USER_TIMESTAMPS % user_name, "%d,%d" % (min_ts, max_ts), USER_CACHE_TIME)
+            if min_ts and max_ts:
+                cache.set(self.ns + REDIS_USER_TIMESTAMPS % user_name, "%d,%d" % (min_ts, max_ts), USER_CACHE_TIME)
 
         return min_ts, max_ts
 
     def insert(self, listens):
-        """ Insert a batch of listens.
+        """
+            Insert a batch of listens. Returns a list of (listened_at, recording_msid, user_name) that indicates
+            which rows were inserted into the DB. If the row is not listed in the return values, it was a duplicate.
         """
 
         submit = []
         user_names = {}
         for listen in listens:
             user_names[listen.user_name] = 1
-            submit.append(listen.to_influx(quote(listen.user_name)))
+            submit.append((listen.ts_since_epoch, listen.recording_msid,
+                           listen.user_name, listen.to_timescale()))
+
+        query = """INSERT INTO listen (listened_at, recording_msid, user_name, data)
+                    VALUES %s
+                    ON CONFLICT (listened_at, recording_msid, user_name)
+                        DO NOTHING
+                        RETURNING listened_at, recording_msid, user_name"""
+
+        inserted_rows = []
+        conn = timescale.engine.raw_connection()
```

Comment:
Yes, because we're using the execute_values method for psycopg2 which allows for more efficient large INSERTs. I think SQLA has support for it, but rather than drinking more SQLA kool-aid, I opted to stick closer to psycopg2.